compiler/rustc_query_system/src/error.rs

```diff
@@ -0,0 +1,60 @@
+use rustc_errors::{DiagnosticBuilder, ErrorGuaranteed};
+use rustc_session::SessionDiagnostic;
+use rustc_span::Span;
+
+pub struct Cycle {
+    pub span: Span,
+    pub stack_bottom: String,
+    pub upper_stack_info: Vec<(Span, String)>,
+    pub recursive_ty_alias: bool,
+    pub recursive_trait_alias: bool,
+    pub cycle_usage: Option<(Span, String)>,
+}
+
+impl SessionDiagnostic<'_> for Cycle {
+    fn into_diagnostic(
+        self,
+        sess: &'_ rustc_session::parse::ParseSess,
+    ) -> DiagnosticBuilder<'_, ErrorGuaranteed> {
+        let mut diag = sess.struct_err(rustc_errors::fluent::query_system::cycle);
+        diag.set_span(self.span);
+        diag.code(rustc_errors::DiagnosticId::Error("E0391".to_string()));
+        let upper_stack_len = self.upper_stack_info.len();
+        for (span, desc) in self.upper_stack_info.into_iter() {
+            // FIXME: use fluent translation
+            diag.span_note(span, &format!("...which requires {}...", desc));
```

Comment:
I wonder if we could add support for `#[subdiagnostic] subs: Vec<SomeSubdiagnosticType>` and represent this that way?